http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2015 Palantir Technologies, Inc. All rights reserved.
+ */
+
+package feign;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableMultimap;
+import com.palantir.remoting.http.FeignClientFactory;
+import com.palantir.remoting.http.ObjectMappers;
+import com.palantir.remoting.http.QueryMap;
+import com.palantir.remoting.http.errors.SerializableErrorErrorDecoder;
+import feign.QueryMapTestServer.TestService;
+import feign.jackson.JacksonDecoder;
+import feign.jackson.JacksonEncoder;
+import feign.jaxrs.JAXRSContract;
+import io.dropwizard.Configuration;
+import io.dropwizard.testing.junit.DropwizardAppRule;
+import javax.net.ssl.SSLSocketFactory;
+import jersey.repackaged.com.google.common.collect.ImmutableList;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+public final class QueryMapUsingServerWithContextTests {
+
+    /**
+     * Creates a {@link QueryMapTestServer.WithContext} Dropwizard server.
+     * The Java implementation of this server uses the context to read all
+     * query parameters from a map and returns them.
+     */
+    @ClassRule
+    public static final DropwizardAppRule<Configuration> APP = new DropwizardAppRule<>(
+            QueryMapTestServer.WithContext.class,
+            "src/test/resources/test-server.yml");
+
+    private QueryMapTestServer.TestClientService service;
+
+    @Before
+    public void before() {
+        String endpointUri = "http://localhost:" + APP.getLocalPort();
+
+        FeignClientFactory factory = FeignClientFactory.of(
+                new JAXRSContract(),
+                // register query map encoder
+                new QueryMapEncoder(new JacksonEncoder(ObjectMappers.guavaJdk7())),
+                new OptionalAwareDecoder(new TextDelegateDecoder(new JacksonDecoder(ObjectMappers.guavaJdk7()))),
```

Comment:
OK cool, yeah, that makes sense. Tagging @uschi2000 and throwing out some ideas that came to my mind (noticed this as well when I was writing this code).

If most of the decoders/encoders are delegating ones, could have a pattern where you just supply them all and have them chained. Should work for most simple cases, although in some cases could lead to some weird dependencies. 

Another approach could be to have some kind of "handler" pattern where the thing you're registering has a predicate to determine whether or not to process the item and will then process it if so. In this scenario, builder/config could take a collection of processors and have a contract that any input will be processed by the first processor whose predicate is satisfied (although this can have corner cases depending on the situation as well like if a handler for a compound type wants to use the "Optional" handler for the sub-components of its type).